File: Leetcode/SortIntegersbyTheNumberof1Bits.rs
// https://leetcode.com/problems/sort-integers-by-the-number-of-1-bits

struct Solution {}

impl Solution {
    pub fn sort_by_bits(mut arr: Vec<i32>) -> Vec<i32> {
        fn bits(mut x: i32) -> usize {
            let mut cnt = 0;
            while x > 0 {
                cnt += 1;
                x &= x - 1;
            }
            cnt
        }

        arr.sort_by(|a, b| match bits(*a).cmp(&bits(*b)) {
            std::cmp::Ordering::Equal => a.cmp(&b),
            r @ _ => r,
        });
        arr
    }
}

File: Leetcode/SortIntegersbyTheNumberof1Bits.cpp
// https://leetcode.com/problems/sort-integers-by-the-number-of-1-bits

class Solution {
   public:
    vector<int> sortByBits(vector<int>& arr) {
        sort(arr.begin(), arr.end(), [](const auto a, const auto b) {
            // return popcount(a) < popcount(b);

            int ac = 0;
            for (auto aa = a; aa; ac += (aa & 1), aa >>= 1) {
            }

            int bc = 0;
            for (auto bb = b; bb; bc += (bb & 1), bb >>= 1) {
            }

            return (ac < bc) || (ac == bc && a < b);
        });

        return arr;
    }
};